### DISCUSSION TITLE
클래스가 관리하는 멤버변수가 많으면 어떤 점이 안좋을까요?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#54] [Clean Code]
"3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다." 라는 요구사항이 왜 있을까?

**Comment by alstn113** [comment]
클래스가 관리하는 멤버 변수가 많을 경우 발생할 수 있는 문제점은 다음과 같습니다.

1. 멤버 변수의 수가 많으면 클래스의 코드가 복잡해지고, 이를 이해하고 유지보수하는 것이 어려워집니다. 
즉, 코드의 가독성이 저하되고, 클래스를 이해하기 어려워집니다.
2. 멤버 변수가 많으면 클래스 간의 의존성이 증가하고, 결합도가 높아집니다. 이는 코드의 유연성을 감소시키고, 재사용성을 저하시킬 수 있습니다.
3. 많은 멤버 변수를 가진 클래스의 버그를 찾고 수정하는 것은 어려울 수 있습니다. 특히 복잡한 상호작용이 있는 경우 디버깅이 더 어려워집니다.

이러한 문제들을 해결하기 위해서는 클래스가 관리하는 멤버 변수의 수를 줄이고, 단일 책임 원칙을 준수하여 클래스를 작게 유지하는 것이 좋습니다. 또한 관련된 데이터를 그룹화하여 적절한 객체로 추출하고, 객체 간의 의존성을 최소화하여 클래스의 복잡성을 줄이는 것이 중요합니다.

**Comment by GIVEN53** [comment]
1. 복잡성 증가
    - 클래스의 복잡성이 증가한다. 클래스를 이해하고 수정하는 것을 어렵게 만들 수 있다.
2. 가독성 저하
    - 코드를 읽기 어렵게 만든다. 클래스의 구조를 이해하기 어려워지며, 코드를 읽는 데 시간이 오래 걸릴 수 있다.
3. 메모리 사용 증가
    - 클래스의 인스턴스가 사용하는 메모리 공간을 증가시킨다. 메모리 사용량을 늘려 성능에 영향을 줄 수 있다.
4. 유지보수 어려움
    - 클래스의 상태를 파악하고 변경 사항을 추적하기 힘들어지기 때문에 유지보수가 어려워진다.
5. 의존성 증가
    - 클래스 간의 의존성을 증가시키기 때문에 클래스를 재사용하거나 테스트하기 어렵게 만들 수 있다.
6. 성능 저하
    - 객체를 생성하고 초기화하는 데 더 많은 시간이 소요될 수 있다. 또한 메모리를 할당하고 해제하는 과정에서 오버헤드가 발생할 수 있다.
